src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -819,26 +826,100 @@ private BoundExpression BindTupleExpression(TupleExpressionSyntax node, Diagnost
                 }
             }
 
+            RemoveDuplicateInferredTupleNames(inferredElementNames, uniqueFieldNames);
             uniqueFieldNames.Free();
-
-            var elementNamesArray = elementNames == null ?
-                                default(ImmutableArray<string>) :
-                                elementNames.ToImmutableAndFree();
+            var (elementNamesArray, inferredArray) = MergeTupleElementNamesAndFree(elementNames, inferredElementNames);
 
             NamedTypeSymbol tupleTypeOpt = null;
             var elements = elementTypes.ToImmutableAndFree();
             var locations = elementLocations.ToImmutableAndFree();
 
             if (hasNaturalType)
             {
-                tupleTypeOpt = TupleTypeSymbol.Create(node.Location, elements, locations, elementNamesArray, this.Compilation, syntax: node, diagnostics: diagnostics, shouldCheckConstraints: true);
+                tupleTypeOpt = TupleTypeSymbol.Create(node.Location, elements, locations, elementNamesArray,
+                    this.Compilation, syntax: node, diagnostics: diagnostics, shouldCheckConstraints: true);
             }
             else
             {
                 TupleTypeSymbol.VerifyTupleTypePresent(elements.Length, node, this.Compilation, diagnostics);
             }
 
-            return new BoundTupleLiteral(node, elementNamesArray, boundArguments.ToImmutableAndFree(), tupleTypeOpt, hasErrors);
+            return new BoundTupleLiteral(node, elementNamesArray, inferredArray, boundArguments.ToImmutableAndFree(), tupleTypeOpt, hasErrors);
+        }
+
+        private static (ImmutableArray<string> names, ImmutableArray<bool> inferred) MergeTupleElementNamesAndFree(
+            ArrayBuilder<string> elementNames, ArrayBuilder<string> inferredElementNames)
+        {
+            if (elementNames == null && inferredElementNames == null)
+            {
+                return (default(ImmutableArray<string>), default(ImmutableArray<bool>));
+            }
+            if (elementNames == null && inferredElementNames != null)
+            {
+                var finalNames = inferredElementNames.ToImmutableAndFree();
+                return (finalNames, finalNames.SelectAsArray(n => n != null));
+            }
```

Comment:
Perhaps nested `if`s for the two `if`s above: `if (elementNames == null) { if (inferredElementNames == null) { ... } else { ... } }`